### PR TITLE
Feat: Add support for change controls (#413)

### LINKED ISSUE #413: Add support for new Change Control Resource APIs
### Enhancement summary

Would be nice to add support to the new CC Resource APIs and add support for 
- multiple stage creations
- ability to add custom script actions and built-in actions to stages
- stage mode definition (parallel vs series)

### Which component of AVD is impacted

cv_task_v3

### Use case example

A few use cases would be to be able to pre-load images or do mlag issu and other things

### Describe the solution you'd like

First cvprac has to support these APIs, which is already WIP
Then we need to add support for image upgrade on the ansible side (that's one of the next things we're working on)
And then we can either build a new module for change controls or update the task module.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by colinmacgiolla**
## Change Summary

This PR adds the ability to get, create/update, and delete change controls.

## Related Issue(s)

Fixes #413

## Component(s) name

`arista.cvp.cv_change_control_v3`

## Proposed changes

The `CvChangeControlTools` class provides a simple module to allow for us to get/set/remove change controls,
```yaml
name: <Name of CC - doesn't have to be unique, but should be to make life easier>
state: <show | set | remove >
change: <dict>
change_id: <list>
```
The `CvpChangeControlBuilder` class does most of the heavy lifting in terms of building the CC data structure, to pass to CVP. The way of describing the change control is relatively simple, but allows for arbitrary (nested) multi-stage changes to be created, to any depth and/or level of complexity required.

```yaml
name: <name of change control>
notes: <Any notes that you want to add>
stages:
   - name: <name of stage>
     mode: <series | parallel>
     parent: <name of parent stage>
activities:
   - name: <only used internally, "task" for any tasks>
     task_id: <str - the WorkOrderId of the task to be executed>
     timeout: <int>
     stage: <str - the name of the Stage to assign the task to>
   - name: <only used internally>
     action: <The name of the action to be done e.g. "Switch Healthcheck">
     stage: <The name of the stage to assign the action to> 
     arguments: <list of dicts, each consisting of a name, and value key>
        - name: <argument name>
          value: <argument value>
```
An example here:
```yaml
name: Ansible playbook test change
notes: Created via playbook
activities:
  - action: "Switch Healthcheck"
    name: Switch1_healthcheck
    arguments:
       - name: DeviceID
         value: SERIAL_NUMBER
    stage: Pre-Checks
  - action: "Switch Healthcheck"
    name: Switch1b_healthcheck
    arguments:
       - name: DeviceID
         value: SERIAL_NUMBER
    stage: Pre-Checks
  - task_id: "20"
    stage: Leaf1a_upgrade
  - task_id: "22"
    stage: Leaf1b_upgrade
stages:
  - name: Pre-Checks
    mode: parallel
  - name: Upgrades
    modes: series
  - name: Leaf1a_upgrade
    parent: Upgrades
  - name: Leaf1b_upgrade
    parent: Upgrades
```
image

## How to test
An ansible test run:
```yaml
---
- name: CVP Change Control Tests
  hosts: cv_server
  gather_facts: no
  vars:
    ansible_command_timeout: 1200
    ansible_connect_timeout: 600
    change:
      name: Ansible playbook test change
      notes: Created via playbook
      activities:
        - action: "Switch Healthcheck"
          name: Switch1_healthcheck
          arguments:
            - name: DeviceID
              value: C13DB880A7DF8A7A5FEB7445C476255C
          stage: Pre-Checks
        - action: "Switch Healthcheck"
          name: Switch1b_healthcheck
          arguments:
            - name: DeviceID
              value: 1FA1BEEE0B875693FE5125374F882EE4
          stage: Pre-Checks
        - task_id: "20"
          stage: Leaf1a_upgrade
        - task_id: "22"
          stage: Leaf1b_upgrade
      stages:
        - name: Pre-Checks
          mode: parallel
        - name: Upgrades
          modes: series
        - name: Leaf1a_upgrade
          parent: Upgrades
        - name: Leaf1b_upgrade
          parent: Upgrades


  tasks:
    - name: "Gather CVP change controls {{inventory_hostname}}"
      arista.cvp.cv_change_control_v3:
        state: show
      register: cv_facts


    - name: "Print out all change controls from {{inventory_hostname}}"
      debug:
        msg: "{{cv_facts}}"


    - name: "Check CC structure"
      debug:
        msg: "{{change}}"


    - name: "Create a change control on {{inventory_hostname}}"
      arista.cvp.cv_change_control_v3:
        state: set
        change: "{{ change }}"
      register: cv_change

    - name: "The Change created has ID {{inventory_hostname}}"
      debug:
        msg: "{{ cv_change }}"

    - name: "Updating Change notes"
      ansible.utils.update_fact:
        updates:
          - path: change.key
            value: "{{ cv_change.data.id }}"
          - path: change.notes
            value: "Change updated"
      register: updated


    - name: "Update the change control on {{inventory_hostname}}"
      arista.cvp.cv_change_control_v3:
        state: set
        change: "{{ updated.change }}"
      register: cv_change


    - name: "Get the updated change control {{inventory_hostname}}"
      arista.cvp.cv_change_control_v3:
        state: show
        name: "{{change.name}}"
      register: cv_facts

    - name: "Show the created CC from {{inventory_hostname}}"
      debug:
        msg: "{{cv_facts}}"


    - name: "Delete the CC from {{inventory_hostname}}"
      arista.cvp.cv_change_control_v3:
        state: remove
        name: "{{change.name}}"
      register: cv_deleted

```

### User Checklist

- N/A

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly. (check the box if not applicable)
- [ ] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

**Comment by sugetha24** [comment]
There seems to be some pep8 and pylint errors that's failing one of the CI tests.

File: ansible_collections/arista/cvp/plugins/module_utils/change_tools.py [hunk 1]
```diff
@@ -0,0 +1,657 @@
...
+                e = "Unable to delete change control. Change name or change_id(s) must be specified"
+                self.__ansible.fail_json(msg="{0}".format(e))
+
+        elif state == "set":
+            changeControl = CvpChangeControlBuilder()
+            changeControl.add_known_uuid([v[1] for v in self.__cc_index])
+            cc_structure = changeControl.build_cc(change, name)
+
+            try:
+                self.__cv_client.post('/api/resources/changecontrol/v1/ChangeControlConfig', data=cc_structure)
```
Comment (noredistribution) [inline ansible_collections/arista/cvp/plugins/module_utils/change_tools.py:687]: shouldn't we use the wrapper here which also has version checks in place?
https://github.com/aristanetworks/cvprac/blob/develop/cvprac/cvp_api.py#L3388
`change_control_create_with_custom_stages(custom_cc=cc_structure)`

File: ansible_collections/arista/cvp/plugins/module_utils/change_tools.py [hunk 2]
```diff
@@ -0,0 +1,657 @@
...
+            A dict describing the request change control
+        """
+        MODULE_LOGGER.debug('Collecting change control: %s', cc_id)
+        if self.apiversion < 3.0:
+            MODULE_LOGGER.debug('Using legacy API call')
+            change = self.__cv_client.api.get_change_control_info(cc_id)
+        else:
+            # Rewrite on cvprac > 1.0.7
+            params = 'key.id={}'.format(cc_id)
+            cc_url = '/api/resources/changecontrol/v1/ChangeControl?' + params
```
Comment (noredistribution) [inline ansible_collections/arista/cvp/plugins/module_utils/change_tools.py:596]: we could use change_control_approval_get_one() here
https://github.com/aristanetworks/cvprac/blob/develop/cvprac/cvp_api.py#L3276

File: ansible_collections/arista/cvp/plugins/module_utils/change_tools.py [hunk 3]
```diff
@@ -0,0 +1,657 @@
...
+        MODULE_LOGGER.debug('Collecting Change controls')
+
+        if self.apiversion < 3.0:
+            MODULE_LOGGER.debug('Trying legacy API call')
+            cc_list = self.__cv_client.api.get_change_controls()
+
+        else:
+            # Rewrite on cvprac > 1.0.7
+            MODULE_LOGGER.debug('Using resource API call')
+            cc_list = self.__cv_client.get('/api/resources/changecontrol/v1/ChangeControl/all')
```
Comment (noredistribution) [inline ansible_collections/arista/cvp/plugins/module_utils/change_tools.py:562]: we could use [change_control_get_all() ](https://github.com/aristanetworks/cvprac/blob/develop/cvprac/cvp_api.py#L3257)

File: ansible_collections/arista/cvp/plugins/module_utils/change_tools.py (hunk 1 above)
Comment (colinmacgiolla) [reply ansible_collections/arista/cvp/plugins/module_utils/change_tools.py:687]: It's only in the dev branch, I'll update to use those when cvprac has them available in a release

File: ansible_collections/arista/cvp/plugins/module_utils/change_tools.py (hunk 2 above)
Comment (colinmacgiolla) [reply ansible_collections/arista/cvp/plugins/module_utils/change_tools.py:596]: It's only in the dev branch, I'll update to use those when cvprac has them available in a release

File: ansible_collections/arista/cvp/plugins/module_utils/change_tools.py (hunk 3 above)
Comment (colinmacgiolla) [reply ansible_collections/arista/cvp/plugins/module_utils/change_tools.py:562]: It's only in the dev branch, I'll update to use those when cvprac has them available in a release

Comment (noredistribution) [reply ansible_collections/arista/cvp/plugins/module_utils/change_tools.py:562]: oh right, forgot we didn't tag the new release yet 😬

**Review by noredistribution — APPROVED** [review]
LGTM!

**Review by noredistribution — APPROVED** [review]
LGTM!

**Comment by sugetha24** [comment]
Hi,
I just noticed that we don't have documentation for this module under docs/v3.

**Comment by colinmacgiolla** [comment]
Will add

**Pull request merged**